lib/pymedphys/_trf/decode/table.py

```diff
@@ -12,197 +12,350 @@
+
+    decoded_rows, column_names = decode_rows(
+        trf_table_contents, version, item_parts_length, item_parts
+    )
 
     table_dataframe = create_dataframe(
         decoded_rows, column_names, CONFIG["time_increment"]
     )
 
-    convert_data_table(
-        table_dataframe, CONFIG["linac_state_codes"], CONFIG["wedge_codes"]
+    table_dataframe = convert_data_table(
+        table_dataframe, CONFIG["linac_state_codes"], CONFIG["wedge_codes"], version
     )
 
     return table_dataframe
 
 
-def decode_data_item(row, group, byteorder) -> int:
-    """Converts the bytes of data items into an integer."""
-    return int.from_bytes(row[group], byteorder=byteorder)
+def decode_rows(trf_table_contents, version, item_parts_length, item_parts):
 
+    column_names_from_dict = CONFIG["item_part_names"]
+    column_names_from_data = [
+        str(item_parts[i]) + "_" + str(item_parts[i + 1])
+        for i in range(0, item_parts_length, 2)
+    ]
 
-def decode_column(raw_table_rows: List[str], column_number: int):
-    """Decode all of the items in a given column."""
-    grouping = 2
-    i = column_number * grouping
-    group = slice(i, i + grouping)
-    byteorder = "little"
-
-    column = np.array(
-        [decode_data_item(row, group, byteorder) for row in raw_table_rows]
-    )
+    # print(CONFIG["version_row"])
+    dtype = CONFIG["version_row"][str(version)]["dtype"]
+    offset = CONFIG["version_row"][str(version)]["offset"]
+    lg_scale = CONFIG["version_row"][str(version)]["lg_scale"]
+    line_grouping = lg_scale * item_parts_length + offset
 
-    return column
+    if version == 1:
 
+        decoded_rows = [
+            np.frombuffer(
+                trf_table_contents[i : i + line_grouping],
+                offset=offset,
+                dtype=np.dtype(dtype),
+            )
+            for i in range(0, len(trf_table_contents), line_grouping)
+        ]
+        column_names = [column_names_from_dict[c] for c in column_names_from_data]
 
-def decode_table_data(raw_table_rows: List[str], line_grouping):
-    """Decode the table into integer values."""
+    else:
 
-    result = []
-    for column_number in range(0, line_grouping // 2):
-        result.append(decode_column(raw_table_rows, column_number))
+        timestamps = [
+            np.frombuffer(trf_table_contents[i : i + 8], offset=0, dtype=np.int64)
+            for i in range(0, len(trf_table_contents), line_grouping)
+        ]
+        item_part_values_data = [
+            np.frombuffer(
+                trf_table_contents[i : i + line_grouping],
+                offset=offset,
+                dtype=np.dtype(dtype),
+            )
+            for i in range(0, len(trf_table_contents), line_grouping)
+        ]
+        decoded_rows = [
+            np.concatenate((timestamps[i], item_parts))
+            for i, item_parts in enumerate(item_part_values_data)
+        ]
+        column_names = ["Timestamp Data"] + [
```

Comment:
this will raise a key not found error when there is a a value c in the column_names_from_data that is not in the column_names_from_dict.  That's appropriate behaviour for python, but...
One could either just add to the dictionary and hope for the best, or
```
column_names_from_dict_including_unknowns = dict(column_names_from_dict)
        for c in column_names_from_data:
            if c not in column_names_from_dict_including_unknowns:
                column_names_from_dict_including_unknowns[c] = "Item: " + c
                print(f'"{c}": "Item: {c}",')
```

and emit the missing items before the exception is raised.
That would let the user/programmer add those items in temporarily for themselves in the config.json file.  Which (looking forward) is what I did (this code is great!) so I could decode some MR Linac TRF data.

Reply:
This is a much better way to deal with and avoid runtime error. Since we have the list for the item parts from the header then it can be completed later.